src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,104 @@
+                await this.CopyFileShare(downloadParameters.ProjectId, downloadRootPath, dirPath, downloadParameters.MinimatchFilters, cancellationToken, isSingleArtifactDownload: false);
+            }
+        }
+
+        public async Task PublishArtifactAsync(string sourcePath, string destPath, int parallelCount)
+        {
+            await this.DirectoryCopyWithMiniMatch(sourcePath, destPath, CancellationToken.None, parallelCount);
+        }
+
+        private async Task CopyFileShare(Guid projectId, string downloadRootPath, string destPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken, bool isSingleArtifactDownload = true)
+        {
+            //minimatchPatterns = minimatchPatterns.Select(pattern => Path.Combine(downloadRootPath, pattern));
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, this.tracer);
+            await DirectoryCopyWithMiniMatch(downloadRootPath, destPath, cancellationToken, defaultParallelCount, minimatcherFuncs);
+        }
+
+        private async Task DirectoryCopyWithMiniMatch(string sourcePath, string destPath, CancellationToken cancellationToken, int parallelCount = defaultParallelCount, IEnumerable<Func<string, bool>> minimatchFuncs = null)
+        {
+            // If the source path is a file, the system should copy the file to the dest directory directly. 
+            if(File.Exists(sourcePath)) {
+                destPath = destPath + Path.DirectorySeparatorChar + Path.GetFileName(sourcePath);
+                this.context.Output(StringUtil.Loc("CopyFileToDestination", sourcePath, destPath));
+                File.Copy(sourcePath, destPath, true);
+                await Task.CompletedTask;
+            }
+
+            // Get the subdirectories for the specified directory.
+            DirectoryInfo dir = new DirectoryInfo(sourcePath);
+
+            // If the destination directory doesn't exist, create it.
+            if (!Directory.Exists(destPath))
+            {
+                Directory.CreateDirectory(destPath);
+            }
+
+            DirectoryInfo[] dirs = dir.GetDirectories();
+                
+            // Get the files in the directory and copy them to the new location.
+            var files = dir.GetFiles("*", SearchOption.AllDirectories);
+
+            var parallelism = new ExecutionDataflowBlockOptions()
+            {
+                MaxDegreeOfParallelism = parallelCount,
+                BoundedCapacity = 2 * parallelCount,
+                CancellationToken = cancellationToken
+            };
+
+            var actionBlock = NonSwallowingActionBlock.Create<FileInfo>(
+                action: file =>
+                {
+                    if (minimatchFuncs == null || minimatchFuncs.Any(match => match(file.FullName))) 
+                    {
+                        string tempPath = Path.Combine(destPath, file.Name);
```

Comment:
this tempPath doesn't seem right:
`string tempPath = Path.Combine(destPath, file.Name);`

By using file.Name, it loses the relative path in the context of the source path.
You need to grab the relative path in context of the source path by using the following:;
https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getrelativepath?view=netstandard-2.1

Else this will just copy the file directly into the destination folder while losing the relative paths.

Reply:
I think you need something like this:
`string tempPath = Path.Combine(destPath, Path.GetRelativePath(sourcePath, file.FullName));`

Double check the logic in a test or manually.